package/kernel/mac80211/files/lib/wifi/mac80211.sh

```diff
@@ -77,9 +93,21 @@ detect_mac80211() {
 		[ "$found" -gt 0 ] && continue
 
 		mode_band="g"
-		channel="11"
+		channel="5"
 		htmode=""
 		ht_capab=""
+		encryption="psk2"
+		key="$secure_psk"
+		[ -n "$wifi_txpower" ] && tx_power="set wireless.radio${devidx}.txpower=$wifi_txpower" \
+			|| [ -n "$wifi_country" ] || tx_power="set wireless.radio${devidx}.txpower=5"
+		[ -n "$wifi_country" ] && country="$wifi_country" || country="XX"
+		[ -n "$wifi_key" ] || bssid="set wireless.default_radio${devidx}.macaddr=$( \
+					dd if=/dev/urandom bs=1 count=6 2>/dev/null \
```

Comment:
This snippet generates multicast MAC addresses which would make this option to fail and thus let an attacker know the correct MAC address which would lead to a security flaw when it's set as PSK.

Reply:
Could you explain this more so we can fix it?

Reply:
There are multicast and unicast MAC addresses. Multicast MAC addresses have the least significant bit of the first byte set to 1. When you set a multicast MAC address for the WLAN interface, it is being ignored silently and the interface still uses the original MAC address which could be identical to the label MAC address or the label MAC address could be derived from it. If the PSK is set to the value of the label MAC address, it's easy for an attacker to get access. Your ash-snippet generates 50% multicast MAC addresses which would mean an attacker has 50% chances to get access and every time the router reboots, the attacker has got another try... You need to make sure that the least significant bit of the first byte of the MAC address is 0. As we're having hexadecimal characters here we might use that, because each character refers to a nibble. So we need to focus on the second character of the MAC address. It must only be 0,2,4,6,8,A,C or E.
Another thing is that as we randomly generate a MAC address here, we should set the "locally administered" bit which is left of the LSB. Thus the second character of the MAC address needs to be 2,6,A or E.

I'm sure you'll find an elegant solution for this.

Reply:
I love your explanation! I´ll fix it ;)

Reply:
You can use the following code to generate a valid locally administered unicast MAC:

```
generate_random_mac() {
	set -- $(hexdump -n 6 -ve '1/1 "0x%x" " "' /dev/urandom)
	printf "%02x:%02x:%02x:%02x:%02x:%02x" \
		$(( ($1 & 0xfe) | 0x02 )) \
		$2 $3 $4 $5 $6
}
```

Reply:
Thank for your suggestion jow! I used to improve the final solution :)